src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokenPsiElement.java

```diff
@@ -0,0 +1,152 @@
+/*******************************************************************************
+ * Copyright (c) 2025 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at http://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+
+package com.redhat.devtools.lsp4ij.features.semanticTokens.viewProvider;
+
+import com.intellij.lang.ASTNode;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiNameIdentifierOwner;
+import com.intellij.psi.impl.source.tree.LeafPsiElement;
+import com.redhat.devtools.lsp4ij.LSPIJUtils;
+import com.redhat.devtools.lsp4ij.features.LSPPsiElement;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+
+/**
+ * A PSI element for a semantic token in a {@link LSPSemanticTokensFileViewProvider} file.
+ */
+public class LSPSemanticTokenPsiElement extends LSPPsiElement implements PsiNameIdentifierOwner {
+    private final LSPSemanticToken semanticToken;
+    private final boolean isFile;
+    private volatile LeafPsiElement node = null;
+    private volatile String text = null;
+
+    /**
+     * Creates a new PSI element for the provided semantic token.
+     *
+     * @param semanticToken the semantic token
+     */
+    LSPSemanticTokenPsiElement(@NotNull LSPSemanticToken semanticToken) {
+        super(semanticToken.getFile(), semanticToken.getTextRange());
+        this.isFile = Objects.equals(semanticToken.getFile().getTextRange(), semanticToken.getTextRange());
```

Comment:
Make note of whether or not this token is for the full file's text range so that we can avoid copying the file's entire text locally in `getText()` below.